File: client/src/data/Stores.js
// HACK - remove before deploy, pass server location from server.
import ChildStore from './ChildStore.js';
import ItemStore from './ItemStore.js';
import SettingsStore from './SettingsStore.js';
import ValuesStore from './ValuesStore.js';

const SERVER_DATA_PREFIX = process.env.REACT_APP_SERVER ? `${process.env.REACT_APP_SERVER}/_` : '/_';

/** Holds all the data within the app. */
const Stores = {
  singleStore:       new ItemStore(`${SERVER_DATA_PREFIX}/single`, 'single'),
  composStore:       new ItemStore(`${SERVER_DATA_PREFIX}/compos`, 'compos'),
  graphsStore:       new ItemStore(`${SERVER_DATA_PREFIX}/graphs`, 'graphs'),
  valuesStore:     new ValuesStore(`${SERVER_DATA_PREFIX}/values`),
  settingsStore: new SettingsStore(`${SERVER_DATA_PREFIX}/settings`),

  // MEGA HACK - assumes compos
  lookUpChildNames(childIds: Array<Object>): Array<Object> {
    const allSingle = this.singleStore.list();
    const allCompos = this.composStore.list();
    if (allSingle === undefined || allCompos === undefined) {
      return undefined;
    }
    return childIds.map(child => {
      let name = undefined;
      if (child.type === "single") {
        const line = allSingle.find(s => s.id === child.id);
        name = line === undefined ? '???' : line.name;
      } else if (child.type === "compos") {
        const line = allCompos.find(s => s.id === child.id);
        name = line === undefined ? '???' : line.name;
      } else if (child.type === "const") {
        name = `Constant (${child.value})`
      }
      return {...child, name};
    });
  }
};

Stores.childStore = new ChildStore(SERVER_DATA_PREFIX, Stores.composStore, Stores.graphsStore);

export default Stores;

File: server/src/java/today/useit/linetracker/store/Stores.java
package today.useit.linetracker.store;

import today.useit.linetracker.model.*;

/** Holder of all stores used within the app. */
public interface Stores {
  /** @return The store for single line metadata. */
  ItemStore<SingleLineMeta> singleStore();

  /** @return The store for composite line metadata. */
  ItemStore<ComposLineMeta> composStore();

  /** @return The store for graph metadata. */
  ItemStore<GraphsLineMeta> graphsStore();

  /** @return Store exposing dated values for single/compos lines. */
  ValuesStore valuesStore();

  /** @return Store mapping parent (compos/graphs) to child (graphs/single). */
  ChildStore childStore();

  /** @return Store holding all per-user settings. */
  SettingsStore settingsStore();
}
